okapi-common/src/main/java/org/folio/okapi/common/ModuleId.java

```diff
@@ -1,34 +1,101 @@
   private final SemVer semVer;
   private final String id;
 
+  static boolean isAsciiLetter(char c) {
+    return c >= 97 && c <= 122; // a-z
+  }
+
+  static boolean isAsciiNumeric(char c) {
+    return c >= 48 && c <= 57;  // 0-9
+  }
+
   /**
-   * Construct Module ID from Module ID string or product only (moduleID without
-   * version).
-   * May throw IllegalArgumentException for invalid syntax for semantic version.
-   * @param s Module ID or product name. The module ID does not have restrictions
-   *     on characters, but a digit following a hyphen-minus marks the beginning of
-   *     a semantic version .. See {@link org.folio.okapi.common.SemVer}.
+   * Construct Module identifier from string.
+   *
+   * <p>The identifier string consists of a product name and optionally followed
+   * by semantic version. Example of valid identifiers:
+   * <ul>
+   *   <li>"foo-1.0.0" : product foo with version 1.0.0.</li>
+   *   <li>"foo-bar-2.0.0-SNAPSHOT" : product foo-bar with version 2.0.0-SNAPSHOT.</li>
+   *   <li>"zoe" : product Zoe without version.</li>
+   *   <li>"folio_email-shipping-1.0.0" : FOLIO UI module.</li>
+   * </ul>
+   * Example of invalid identifiers:
+   * <ul>
+   *   <li>"mod_inventory"</li>
+   *   <li>"FOO"</li>
+   *   <li>"foo 1.0.0"</li>
+   *   <li>"123"</li>
+   *   <li>"a/b"</li>
+   * </ul>
+   * @param s Module ID or product name. The module ID may include of lowercase letters, digits,
+   *          and hyphen-minus. Leading character, however, must be lower-case letter.
+   *          A digit following a hyphen-minus marks the beginning of a semantic version.
+   *          See {@link org.folio.okapi.common.SemVer}.
+   * @throws IllegalArgumentException if the module identifier is invalid.
    */
   public ModuleId(String s) {
     id = s;
-    for (int i = 0; i < s.length() - 1; i++) {
-      if (s.charAt(i) == '-' && Character.isDigit(s.charAt(i + 1))) {
-        product = s.substring(0, i);
-        semVer = new SemVer(s.substring(i + 1));
-        return;
+    if (s.isEmpty()) {
+      throw new IllegalArgumentException("ModuleID must not be empty");
+    }
+    if (!isAsciiLetter(s.charAt(0))) {
+      throw new IllegalArgumentException(MODULE_QUOTE + id + "' must start with lowercase letter");
+    }
+    final boolean ui = s.startsWith("folio_");
+    int i = ui ? 6 : 1;
+    for (; i < s.length(); i++) {
+      char c = s.charAt(i);
+      if (c == '-') {
+        if (i >= s.length() - 1) {
+          throw new IllegalArgumentException(MODULE_QUOTE + id
+              + "' has hyphen-minus as last character");
+        }
+        c = s.charAt(++i);
+        if (isAsciiNumeric(c)) {
+          product = s.substring(0, i - 1);
+          productCheck(ui);
+          semVer = new SemVer(s.substring(i));
+          return;
+        }
+        if (!isAsciiLetter(c)) {
+          throw new IllegalArgumentException(MODULE_QUOTE + id
+              + "' has non-allowed character at offset " + i);
```

Comment:
1 based position is more readable than 0 based offset.
`at position " + (i + 1)`